### PR TITLE
handle terminal switch() in return_linter()

### LINKED ISSUE #2343: return_linter() should allow switch() where every entry has an exit call
@MichaelChirico I don't think we should give a lint here. In my opinion this is a false positive, as there is no way an implicit return is given.
My first implementation handled this accordingly. Maybe we could do that again.

_Originally posted by @MEO265 in https://github.com/r-lib/lintr/pull/2271#discussion_r1401041242_

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #2343.

cc @MEO265

### CODE REVIEW

**Comment by MichaelChirico** [comment]
😆 Definitely my most unexpected time getting hit with the `commented_code_linter()` 🔨

```r
switch(x, ...) | expr[1]: switch; expr[2]: x
```

Pretty impressive that parses as valid code!

**Comment by AshesITR** [comment]
> 😆 Definitely my most unexpected time getting hit with the `commented_code_linter()` 🔨
> 
> ```r
> switch(x, ...) | expr[1]: switch; expr[2]: x
> ```
> 
> Pretty impressive that parses as valid code!

I like to prepend the code by `#>`, which generally kills parsing due to a missing LHS of `>` ;)

**Comment by MichaelChirico** [comment]
> > 😆 Definitely my most unexpected time getting hit with the `commented_code_linter()` 🔨
> > ```r
> > switch(x, ...) | expr[1]: switch; expr[2]: x
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Pretty impressive that parses as valid code!
> 
> I like to prepend the code by `#>`, which generally kills parsing due to a missing LHS of `>` ;)

right, but I would reserve that for code that's actually intended as code. In this case, I was writing a comment to explain the XPath

File: tests/testthat/test-return_linter.R [hunk 1]
```diff
@@ -1521,3 +1471,231 @@ test_that("= assignments are handled correctly", {
...
+  all_return_lines <- trim_some("
+    foo <- function(x) {
+      switch(x,
+        a = return(1),
+        b = return(2)
+      )
+    }
+  ")
+  expect_lint(all_return_lines, list(implicit_msg, implicit_msg), implicit_linter)
+  expect_lint(all_return_lines, NULL, explicit_linter)
```
Comment (AshesITR) [inline tests/testthat/test-return_linter.R:1523]: Should the behavior of `explicit_linter` here depend on `allow_implicit_else`, i.e. force a `return()` in the default case if the latter is `FALSE`?

Comment (MichaelChirico) [reply tests/testthat/test-return_linter.R:1523]: interesting Q... I lean towards yes. @MEO265 any thoughts?

Comment (AshesITR) [reply tests/testthat/test-return_linter.R:1523]: Ditto actually for implicit return style requiring an unnamed arg to `switch()`.

Comment (MichaelChirico) [reply tests/testthat/test-return_linter.R:1589]: OK, `allow_implicit_else` also covers whether the default case is required in a terminal `switch()` statement.

**Review by AshesITR — APPROVED** [review]
Awesome!

**Review by AshesITR — APPROVED** [review]

**Comment by MichaelChirico** [comment]
roxygen test handy already 🤭

**Review by AshesITR — APPROVED** [review]

**Pull request merged**